File: clojure/beer-song/src/beer_song.clj
(ns beer-song
  (:require [clojure.string :refer [join]]))

(defn bottles [n]
  (case n
      1 "1 bottle"
      0 "no more bottles"
      (str n " bottles")))

(defn capital-bottles [n]
  (if (= 0 n) "No more bottles"
      (bottles n)))

(defn one [n]
  (if (= 1 n) "it"
      "one"))

(defn second-line [v]
  (str
   (if (= 0 v) "Go to the store and buy some more, "
       (str "Take " (one v) " down and pass it around, "))
   (bottles (mod (dec  v) 100)) " of beer on the wall.\n"))

(defn verse [v]
  (str (capital-bottles v) " of beer on the wall, " (bottles v) " of beer.\n"
       (second-line v)))

(defn sing
  ([start-verse] (sing start-verse 0))
  ([start end] (join "\n" (map verse (range start (dec end) -1)))))

File: ecmascript/beer-song/beer-song.js
const bottles =
      no => no == 0 ? "no more bottles"
      : no == 1 ? "1 bottle"
      : `${no} bottles`

const capitalBottles =
      no => no == 0 ? "No more bottles"
      : bottles(no)

const takeIt =
      no => no == 0 ? "Go to the store and buy some more"
      : no == 1 ? "Take it down and pass it around"
      : "Take one down and pass it around"

const range = (start, end) => [...Array(end - start + 1).keys()]
      .map(el => el + start)

export default {
  verse(no) {
    return capitalBottles(no) + " of beer on the wall, "
      + bottles(no) + " of beer.\n" + takeIt(no) + ", "
      + bottles((no + 99) % 100) + " of beer on the wall.\n"
  },

  sing(from = 99, to = 0) {
    return range(to, from).reverse().map(no => this.verse(no)).join('\n')
  }
}
